### PR TITLE
Use MI to call Maestro from PCS

### PR DESCRIPTION
**Pull request by premun**
https://dev.azure.com/dnceng/internal/_workitems/edit/6451

### Release Note Category
- [x] Feature changes/additions 
- [ ] Bug fixes
- [ ] Internal Infrastructure Improvements

### Release Note Description
Enable managed identity based auth for PCS->Maestro calls

### CODE REVIEW

File: src/Maestro/Client/src/MaestroApiCredential.cs [hunk 1]
```diff
@@ -87,6 +87,24 @@ internal static MaestroApiCredential CreateFederatedCredential(string barApiBase
...
         }
 
+        /// <summary>
+        /// Use this for darc invocations from services using an MI
+        /// </summary>
+        internal static MaestroApiCredential CreateManagedIdentityCredential(string barApiBaseUri, string managedIdentityId)
+        {
+            string appId = EntraAppIds[barApiBaseUri.TrimEnd('/')];
+
+            var miCredential = new ManagedIdentityCredential(managedIdentityId);
```
Comment (dkurepa) [inline src/Maestro/Client/src/MaestroApiCredential.cs:97]:
```suggestion
            ManagedIdentityCredential miCredential = new(managedIdentityId);
```
aren't we trying to use the new syntax everywhere?

File: src/ProductConstructionService/ProductConstructionService.Client/Generated/ProductConstructionServiceApi.cs [hunk 2]
```diff
@@ -38,7 +38,7 @@ public partial interface IServiceOperations<T>
     public partial class ProductConstructionServiceApiOptions : ClientOptions
     {
         public ProductConstructionServiceApiOptions()
-            : this(new Uri("https://product-construction-int.wittytree-28a89311.westus2.azurecontainerapps.io/"))
+            : this(new Uri("https://product-construction-int.delightfuldune-c0f01ab0.westus2.azurecontainerapps.io/"))
```
[the remark this answers is not in the document]
Comment (premun) [reply src/ProductConstructionService/ProductConstructionService.Client/Generated/ProductConstructionServiceApi.cs:41]: Done

File: src/ProductConstructionService/ProductConstructionService.Api/Configuration/PcsConfiguration.cs [hunk 3]
```diff
@@ -70,18 +70,26 @@ public static string GetRequiredValue(this IConfiguration configuration, string
...
 
-            // TODO https://dev.azure.com/dnceng/internal/_workitems/edit/6451: Implement Maestro - PCS communication
-            var token = config.GetValue<string>("Maestro:Token");
+            var noAuth = config.GetValue<bool>("Maestro:NoAuth");
+            if (noAuth)
+            {
+                return MaestroApiFactory.GetAnonymous(uri);
+            }
+
+            var managedIdentityId = config.GetValue<string>("ManagedIdentityClientId");
```
Comment (dkurepa) [inline src/ProductConstructionService/ProductConstructionService.Api/Configuration/PcsConfiguration.cs:85]:
```suggestion
            var managedIdentityId = config[ManagedIdentityId];
```
we have a const with the key value

File: src/ProductConstructionService/ProductConstructionService.Api/Configuration/PcsConfiguration.cs [hunk 4]
```diff
@@ -70,18 +70,25 @@ public static string GetRequiredValue(this IConfiguration configuration, string
         builder.AddGitHubClientFactory();
         builder.AddWorkitemQueues(azureCredential, waitForInitialization: initializeService);
 
-        builder.Services.AddSingleton<IMaestroApi>(s =>
+        builder.Services.AddScoped<IMaestroApi>(s =>
         {
-            var config = s.GetRequiredService<IConfiguration>();
-            var uri = config.GetValue<string>("Maestro:Uri")
+            var uri = builder.Configuration.GetValue<string>("Maestro:Uri")
```
Comment (dkurepa) [inline src/ProductConstructionService/ProductConstructionService.Api/Configuration/PcsConfiguration.cs:75]: are there any benefits in using `Configuration.GetValue<string>("..")` over `Configuration[".."]?

Comment (dkurepa) [reply src/ProductConstructionService/ProductConstructionService.Api/Configuration/PcsConfiguration.cs:75]: In the `bool` case we're avoiding the casting, but for strings it seems shorter

**Review by dkurepa — APPROVED** [review]

**Pull request merged**